### PR TITLE
Add operation metadata metrics publishing

### PR DESCRIPTION
**Pull request by 80degreeswest**

### CODE REVIEW

File: src/main/java/build/buildfarm/server/BuildFarmServer.java [hunk 1]
```diff
@@ -116,6 +120,17 @@ private static BuildFarmServerConfig toBuildFarmServerConfig(
...
   }
 
+  private static MetricsPublisher getMetricsPublisher(MetricsConfig metricsConfig) {
+    switch (metricsConfig.getMetricsDestination()) {
+      default:
+        return new LogMetricsPublisher(metricsConfig);
+      case "aws":
+        return new AwsMetricsPublisher(metricsConfig);
+      case "gcp":
+        throw new UnsupportedOperationException("Not Implemented.");
```
Comment (werkt) [inline src/main/java/build/buildfarm/server/BuildFarmServer.java:130]: Let the implementation throw on bad usage, you've provided the class already

File: src/main/java/build/buildfarm/BUILD [hunk 2]
```diff
@@ -74,6 +74,77 @@ java_library(
...
+    name = "common-metrics",
+    srcs = glob([
+        "common/metrics/AbstractMetricsPublisher.java",
+        "common/metrics/MetricsPublisher.java",
+    ]),
+    deps = [
+        "@maven//:com_google_guava_guava",
+        "@maven//:com_google_protobuf_protobuf_java",
+        "@maven//:com_google_protobuf_protobuf_java_util",
+        "//src/main/protobuf:build_buildfarm_v1test_buildfarm_java_proto",
```
Comment (werkt) [inline src/main/java/build/buildfarm/BUILD:87]: deps ordering here and below.

File: src/main/java/build/buildfarm/common/metrics/aws/AwsMetricsPublisher.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    } catch (Exception e) {
+      logger.log(Level.WARNING, String.format("Could not publish request metadata to SNS for %s.", operation.getName()), e);
+    }
+  }
+
+  private AmazonSNSAsync initSnsClient() {
+    logger.log(Level.INFO, "Initializing SNS Client.");
+    return
+      AmazonSNSAsyncClientBuilder.standard()
+        .withRegion(Regions.US_EAST_1)
```
Comment (werkt) [inline src/main/java/build/buildfarm/common/metrics/aws/AwsMetricsPublisher.java:83]: This needs to be config'd

File: src/main/java/build/buildfarm/BUILD (hunk 2 above)
Comment (80degreeswest) [reply src/main/java/build/buildfarm/BUILD]: We should look at the rest of the build files as well as the same ordering exists in other places.

File: src/main/java/build/buildfarm/common/metrics/gcp/GcpMetricsPublisher.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+public class GcpMetricsPublisher extends AbstractMetricsPublisher {
+  private static final Logger logger = Logger.getLogger(GcpMetricsPublisher.class.getName());
+
+  public GcpMetricsPublisher(MetricsConfig metricsConfig) {
+    super(metricsConfig.getClusterId());
+  }
+
+  public GcpMetricsPublisher() {
+    super();
```
Comment (werkt) [inline src/main/java/build/buildfarm/common/metrics/gcp/GcpMetricsPublisher.java:29]: this super call shouldn't be necessary

File: src/main/java/build/buildfarm/common/metrics/log/LogMetricsPublisher.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    super(metricsConfig.getClusterId());
+    if (!metricsConfig.getLogMetricsConfig().getLogLevel().isEmpty()) {
+      logLevel = Level.parse(metricsConfig.getLogMetricsConfig().getLogLevel());
+    } else {
+      logLevel = Level.FINEST;
+    }
+  }
+
+  public LogMetricsPublisher() {
+    super();
```
Comment (werkt) [inline src/main/java/build/buildfarm/common/metrics/log/LogMetricsPublisher.java:39]: Shouldn't be necessary

**Pull request merged**